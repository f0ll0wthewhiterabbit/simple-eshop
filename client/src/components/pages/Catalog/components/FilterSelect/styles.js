import styled from 'styled-components'
import { FormControl } from '@material-ui/core'

export const Wrapper = styled.div`
  position: absolute;
  top: 49px;
  right: 32px;
`

export const SelectFormControl = styled(FormControl)`
  min-width: 60px;
  margin-left: 10px;

  .MuiInputBase-root {
    font-size: 1.2rem;
  }

  .MuiSelect-root {
    padding-left: 8px;
  }
`
