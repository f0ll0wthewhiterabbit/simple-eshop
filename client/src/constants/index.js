export * from './pagePaths'
export * from './endpoints'
export * from './fields'
export * from './roles'
export * from './pageLimits'
