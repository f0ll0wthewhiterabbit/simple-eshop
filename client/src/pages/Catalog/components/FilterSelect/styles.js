import styled from 'styled-components'
import { FormControl } from '@material-ui/core'

export const Wrapper = styled.div`
  text-align: center;
  margin-bottom: 28px;

  ${props => props.theme.breakpoints.up('sm')} {
    position: absolute;
    top: 49px;
    right: 32px;
    margin-bottom: 0;
  }
`

export const SelectFormControl = styled(FormControl)`
  min-width: 60px;
  margin-left: 10px;

  .MuiInputBase-root {
    font-size: 1.2rem;
  }

  .MuiSelect-root {
    padding-left: 8px;
  }
`
