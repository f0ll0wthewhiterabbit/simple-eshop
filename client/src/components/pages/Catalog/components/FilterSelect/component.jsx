import React from 'react'
import PropTypes from 'prop-types'
import { Select, MenuItem } from '@material-ui/core'

import { Wrapper, SelectFormControl } from './styles'
import { URL_FIELD_NO_FILTER, URL_FIELD_RATINGS_FILTER } from '../../../../../constants'

const FilterSelect = ({ filter, fetchProducts, setProductsFilter }) => {
  const handleChange = event => {
    const { value } = event.target
    if (value === filter) {
      return
    }

    setProductsFilter(value)

    fetchProducts(1, null, value)
  }

  return (
    <Wrapper>
      <SelectFormControl>
        <Select value={filter} onChange={handleChange} data-test="select">
          <MenuItem value={URL_FIELD_NO_FILTER}>All</MenuItem>
          <MenuItem value={URL_FIELD_RATINGS_FILTER}>My ratings</MenuItem>
        </Select>
      </SelectFormControl>
    </Wrapper>
  )
}

FilterSelect.propTypes = {
  filter: PropTypes.string.isRequired,
  fetchProducts: PropTypes.func.isRequired,
  setProductsFilter: PropTypes.func.isRequired,
}

export default FilterSelect
