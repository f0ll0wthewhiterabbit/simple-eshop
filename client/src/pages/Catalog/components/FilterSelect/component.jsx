import React from 'react'
import PropTypes from 'prop-types'
import { Select } from '@material-ui/core'
import ExpandMoreIcon from '@material-ui/icons/ExpandMore'

import { URL_FIELD_NO_FILTER, URL_FIELD_RATINGS_FILTER } from '../../../../constants'
import { Wrapper, SelectFormControl, SelectItem } from './styles'

const FilterSelect = ({ filter, isDarkTheme, fetchProducts, setProductsFilter }) => {
  const listProps = isDarkTheme ? { style: { backgroundColor: '#687579' } } : {}

  const handleChange = event => {
    const { value } = event.target
    if (value === filter) {
      return
    }

    setProductsFilter(value)

    fetchProducts(1, null, value)
  }

  return (
    <Wrapper>
      <SelectFormControl>
        <Select
          value={filter}
          onChange={handleChange}
          IconComponent={ExpandMoreIcon}
          data-test="select"
          MenuProps={{
            anchorOrigin: {
              vertical: 'bottom',
              horizontal: 'left',
            },
            transformOrigin: {
              vertical: 'top',
              horizontal: 'left',
            },
            getContentAnchorEl: null,
            MenuListProps: listProps,
          }}
        >
          <SelectItem value={URL_FIELD_NO_FILTER}>All</SelectItem>
          <SelectItem value={URL_FIELD_RATINGS_FILTER}>My ratings</SelectItem>
        </Select>
      </SelectFormControl>
    </Wrapper>
  )
}

FilterSelect.propTypes = {
  filter: PropTypes.string.isRequired,
  isDarkTheme: PropTypes.bool.isRequired,
  fetchProducts: PropTypes.func.isRequired,
  setProductsFilter: PropTypes.func.isRequired,
}

export default FilterSelect
