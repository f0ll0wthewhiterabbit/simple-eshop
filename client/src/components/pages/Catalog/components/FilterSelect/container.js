import { connect } from 'react-redux'

import FilterSelect from './component'
import { fetchProducts, setProductsFilter } from '../../../../../store/actions'

const mapStateToProps = state => ({
  filter: state.getIn(['products', 'filter']),
})

const mapDispatchToProps = dispatch => ({
  fetchProducts: (page, itemsPerPage, filter) =>
    dispatch(fetchProducts(page, itemsPerPage, filter)),
  setProductsFilter: filter => dispatch(setProductsFilter(filter)),
})

export default connect(mapStateToProps, mapDispatchToProps)(FilterSelect)
