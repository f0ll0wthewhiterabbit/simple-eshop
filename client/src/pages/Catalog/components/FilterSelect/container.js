import { connect } from 'react-redux'

import FilterSelect from './component'
import { fetchProducts, setProductsFilter } from '../../../../store/actions'
import { FIELD_THEME_DARK } from '../../../../constants'

const mapStateToProps = state => ({
  filter: state.getIn(['products', 'filter']),
  isDarkTheme: state.getIn(['app', 'theme']) === FIELD_THEME_DARK,
})

const mapDispatchToProps = {
  fetchProducts: (page, itemsPerPage, filter) => fetchProducts(page, itemsPerPage, filter),
  setProductsFilter: filter => setProductsFilter(filter),
}

export default connect(mapStateToProps, mapDispatchToProps)(FilterSelect)
