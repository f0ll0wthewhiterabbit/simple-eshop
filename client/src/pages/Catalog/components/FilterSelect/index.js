import FilterSelect from './container'

export default FilterSelect
