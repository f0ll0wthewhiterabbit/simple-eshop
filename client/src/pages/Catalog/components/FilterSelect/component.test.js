import React from 'react'
import { shallow } from 'enzyme'

import FilterSelect from './component'
import findByTestAttr from '../../../../utils/findByTestAttr'
import { FIELDS } from '../../../../constants'

describe('FilterSelect component', () => {
  let wrapper
  const initialProps = {
    filter: FIELDS.URL_NO_FILTER,
    isDarkTheme: false,
    fetchProductsRequest: jest.fn(),
    setProductsFilter: jest.fn(),
  }
  const dataTestSelect = 'select'

  const generateWrapper = passedProps => {
    const defaultProps = { ...initialProps }
    const props = { ...defaultProps, ...passedProps }

    return shallow(<FilterSelect {...props} />)
  }

  beforeEach(() => {
    wrapper = generateWrapper()
  })

  afterEach(() => {
    initialProps.fetchProductsRequest.mockClear()
    initialProps.setProductsFilter.mockClear()
  })

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should render filter from props', () => {
    const select = findByTestAttr(wrapper, dataTestSelect)

    expect(select.props().value).toBe(FIELDS.URL_NO_FILTER)
  })

  it('should handle filter change', () => {
    const select = findByTestAttr(wrapper, dataTestSelect)
    select.simulate('change', { target: { value: FIELDS.URL_RATINGS_FILTER } })

    expect(initialProps.setProductsFilter).toHaveBeenCalledTimes(1)
    expect(initialProps.setProductsFilter).toHaveBeenCalledWith(FIELDS.URL_RATINGS_FILTER)
    expect(initialProps.fetchProductsRequest).toHaveBeenCalledTimes(1)
    expect(initialProps.fetchProductsRequest).toHaveBeenCalledWith(
      1,
      null,
      undefined,
      FIELDS.URL_RATINGS_FILTER
    )
  })

  it('should change select menu background color in dark mode', () => {
    wrapper = generateWrapper({ isDarkTheme: true })

    expect(wrapper).toMatchSnapshot()
  })
})
