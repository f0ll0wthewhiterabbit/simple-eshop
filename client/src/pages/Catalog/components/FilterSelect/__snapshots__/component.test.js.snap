// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSelect component should change select menu background color in dark mode 1`] = `
<styled.div>
  <Styled(WithStyles(ForwardRef(FormControl)))>
    <WithStyles(ForwardRef(Select))
      IconComponent={
        Object {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "displayName": "ExpandMoreIcon",
          "muiName": "SvgIcon",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        }
      }
      MenuProps={
        Object {
          "MenuListProps": Object {
            "style": Object {
              "backgroundColor": "#687579",
            },
          },
          "anchorOrigin": Object {
            "horizontal": "left",
            "vertical": "bottom",
          },
          "getContentAnchorEl": null,
          "transformOrigin": Object {
            "horizontal": "left",
            "vertical": "top",
          },
        }
      }
      data-test="select"
      onChange={[Function]}
      value="all"
    >
      <Styled(WithStyles(ForwardRef(MenuItem)))
        value="all"
      >
        All
      </Styled(WithStyles(ForwardRef(MenuItem)))>
      <Styled(WithStyles(ForwardRef(MenuItem)))
        value="myRatings"
      >
        My ratings
      </Styled(WithStyles(ForwardRef(MenuItem)))>
    </WithStyles(ForwardRef(Select))>
  </Styled(WithStyles(ForwardRef(FormControl)))>
</styled.div>
`;

exports[`FilterSelect component should render correctly 1`] = `
<styled.div>
  <Styled(WithStyles(ForwardRef(FormControl)))>
    <WithStyles(ForwardRef(Select))
      IconComponent={
        Object {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "displayName": "ExpandMoreIcon",
          "muiName": "SvgIcon",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        }
      }
      MenuProps={
        Object {
          "MenuListProps": Object {},
          "anchorOrigin": Object {
            "horizontal": "left",
            "vertical": "bottom",
          },
          "getContentAnchorEl": null,
          "transformOrigin": Object {
            "horizontal": "left",
            "vertical": "top",
          },
        }
      }
      data-test="select"
      onChange={[Function]}
      value="all"
    >
      <Styled(WithStyles(ForwardRef(MenuItem)))
        value="all"
      >
        All
      </Styled(WithStyles(ForwardRef(MenuItem)))>
      <Styled(WithStyles(ForwardRef(MenuItem)))
        value="myRatings"
      >
        My ratings
      </Styled(WithStyles(ForwardRef(MenuItem)))>
    </WithStyles(ForwardRef(Select))>
  </Styled(WithStyles(ForwardRef(FormControl)))>
</styled.div>
`;
